packages/core-data/src/selectors.ts

```diff
@@ -1145,10 +1145,17 @@ export function isPreviewEmbedFallback( state: State, url: string ): boolean {
 export function canUser(
 	state: State,
 	action: string,
-	resource: string,
+	resource: string | Record< string, any >,
 	id?: EntityRecordKey
 ): boolean | undefined {
-	const key = [ action, resource, id ].filter( Boolean ).join( '/' );
+	const key = (
+		typeof resource === 'object'
+			? [ action, resource.kind, resource.name, resource.id ]
+			: [ action, resource, id ]
```

Comment:
Let's do some consistency checks on the `resource` object, and also prevent conflicting keys. The following calls (some of them buggy) lead to the same `key`:
```js
canUser( 'update', 'posts' );
canUser( 'update', { id: 'posts' } );
canUser( 'update', { name: 'posts' } );
```
The key should have some prefix (`ent:postType/post/1`, `res:posts/1`) and also we should trigger failure if there is `resource.id` and no `resource.name` etc.

Reply:
Where is there a benefit for strings? It doesn't look like it's needed for memoization?

Reply:
I'm talking about the `key` that is used for lookup in Redux state: `state.userPermissions[ key ]`. That is a string. The `resource` parameter for `canUser` is another beast.

Reply:
I guess we can return `null` from the selector when the resource is an object, but the `kind` and `name` properties are missing. These two are required to make any entity queries. Resolve can bail early and not make a request.

I can add a warning or error, but these are rare in WP stores.

Reply:
Yes, that's what I meant by "failure", simply returning `null` 🙂 All we need is to avoid storing or returning the wrong data.

Reply:
Updated in 19e01ea3ac24e542514ddae27addbfaabcff648b.

Reply:
One more idea: can we return `false` when the entity is malformed? With that we can avoid making the return type too complicated with `boolean | undefined | null`. And `false` is still a good answer. `canUser` perform some ill-defined operation? The answer is no.

Reply:
I guess we can throw an error when the resource object is malformed, and the error will be accessible via `getResolutionError`. We're already doing something similar when action isn't supported.

I don't know the historical reason why resolvers/selectors don't log errors; my best guess is that people often report them as bugs. So basically, support overhead.

cc @youknowriad

Reply:
I'm not very happy about throwing an error from the selector. It's not forward compatible. It's likely that we will add a new resource type in the future (e.g., for checking capabilities on the site) and then the new code will be failing on an older WordPress. It's better to return `false`, that means you cannot do the thing you ask for on an older WP version, and you're feature is effectively disabled. Let's just log a warning to the console, so that the error is not completely swallowed.

Reply:
The error is thrown from resolver b1e5dfc2ee25026e0892aa2a79525dd2e7d42bf3.

I think returning `false` is the correct approach. Core does the same with `current_user_can`. Calling `current_user_can( 'fly' )` will return `false`. It's up to the devs to pass the correct data. Otherwise, we'll have to pollute the codebase with warnings.

Fun fact: Calling `current_user_can( 'exist' )` will always return `true`.

Reply:
> The error is thrown from resolver

Oh, I missed that, in resolver it's fine.

> Fun fact: Calling `current_user_can( 'exist' )` will always return true.

It was added in 2010 here: https://core.trac.wordpress.org/ticket/14696#comment:5

> Add an 'exist' capability that all users have, even if they don't have a role on a blog. current_user_can('exist'), yeah, cute. Suggestions welcome.